internals/plan/extensions_test.go

```diff
@@ -0,0 +1,680 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	. "gopkg.in/check.v1"
+	"gopkg.in/yaml.v3"
+
+	"github.com/canonical/pebble/internals/plan"
+)
+
+type inputLayer struct {
+	order int
+	label string
+	yaml  string
+}
+
+// PlanResult represents the final content of a combined plan. Since this
+// package exclusively focuses extensions, all built-in sections are empty
+// and ignored in the test results.
+type planResult struct {
+	x *xSection
+	y *ySection
+}
+
+var extensionTests = []struct {
+	extensions             map[string]plan.LayerSectionExtension
+	layers                 []*inputLayer
+	errorString            string
+	combinedPlanResult     *planResult
+	combinedPlanResultYaml string
+}{
+	// Index 0: No Sections
```

Comment:
Instead of using indexes here, what about a test summary field like `plan_test` does? Like `summary: "No Sections"`. Seems a bit more human-friendly.

Reply:
This is just a personal preference, but I am happy to change it. This is how I use it:

If I see an unit test failure, the last message contains an index, e.g:

```
Running TestPlanExtensions with test data index 3"
```

All I have to do now is append the index to the unit test data set:

```
var extensionTests = []struct {
    extensions map[string]plan.LayerSectionExtension
    layers     []*inputLayer
    error      string
    result     *planResult
    resultYaml string
}{
  : // lots and lots of test cases
}[3:]  // <=======
```

and when I delve debug the issue, I get to immediately debug the failed test.

```
dlv exec plan.test -- -check.f TestPlanExtensions
> break TestPlanExtensions
```

So both the test name, and data index logged helps me to immediately debug the issue.

Without an index, the offset within the test data set is not immediately available.

Let me know if this is a good enough reason to try something different. Happy to revert back to a description only apporach.

Reply:
Interesting approach. I'd probably a conditional breakpoint for this. I use GoLand, though it looks with Delve you can set a [condition](https://github.com/go-delve/delve/tree/master/Documentation/cli#condition) on your breakpoints too: `condition <breakpoint name or id> test.summary == "Foo bar"`.

Or if you're changing your code anyway, you could add an `if test.summary != "Foo Bar" { continue }` line in the test loop.

If the above helps, great, let's use the non-index approach. But if you'd prefer to leave it as is, I don't have strong feelings.

Reply:
I've changed to the summary approach, but the message printed by each test (visible on failure) shows:

```c.Logf("TestPlanExtensions :: %s (data index %v)", testData.summary, testIndex)```

So now its the same as plan_test, but I do still add a hint on the data index. This is more consistent, thanks.

Reply:
Resolving.